sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/planning/ScanOperationSuite.scala

```diff
@@ -73,17 +80,20 @@ class ScanOperationSuite extends SparkFunSuite {
 
   test("Filter which has the same non-deterministic expression with its child Project") {
     val filter1 = Filter(EqualTo(colR, Literal(1)), Project(Seq(colA, aliasR), relation))
-    assert(ScanOperation.unapply(filter1).isEmpty)
+    filter1 match {
+      case ScanOperation(projects, filters, _: Filter) =>
+        assert(projects.size === 2)
+        assert(filters.isEmpty)
+      case _ => assert(false)
+    }
   }
 
   test("Deterministic filter with a child Project with a non-deterministic expression") {
     val filter2 = Filter(EqualTo(colA, Literal(1)), Project(Seq(colA, aliasR), relation))
     filter2 match {
-      case ScanOperation(projects, filters, _: LocalRelation) =>
+      case ScanOperation(projects, filters, _: Filter) =>
```

Comment:
@viirya this is a real change. After the pattern match, the caller side will put Filter first then Project, which means the Filter is pushed through the Project. According to the optimizer rule, this pushdown should not happen if Project has nondeterministic expressions.

Reply:
Above is for previous behavior, right? So it sounds like a (possible) bug?

Reply:
Yea, let me pull out this bug fix